File: valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        '''
        Determines if a Sudoku doesn't break any basic Sudoku rules
        Returns true if Sudoku is valid, otherwise false
        '''
        #Setup a bit mask to keep track of seen values
        rows = [0] * 9
        columns = [0] * 9
        boxes = [0] * 9

        for r in range(0,9): #Iterate rows
            for c in range(0,9): #Iterate columns
                if board[r][c] != '.': #Only compare filled elements
                    current = int(board[r][c]) - 1

                    if rows[r] & (1 << current): #Compare if element present in row
                        return False
                    rows[r] |= (1 << current) #Push bit into list

                    if columns[c] & (1 << current): #Compare if element present in column
                        return False
                    columns[c] |= (1 << current)

                    b_loc = (r // 3) * 3 + c // 3 #Determine box location
                    if boxes[b_loc] & (1 << current): #Compare if element present in box
                        return False
                    boxes[b_loc] |= (1 << current)
        return True

File: valid-sudoku/valid-sudoku.java
class Solution 
{
    public boolean isValidSudoku(char[][] board) 
    {
        return rowValids(board) && colValids(board) && squareValids(board);
    }
    
    /* Ensure rows are valid */
    public boolean rowValids(char[][] board)
    {
        for(int row = 0; row < board.length; row++)
        {
            HashSet<Character> rows = new HashSet<Character>();
            for(int col = 0; col < 9; col++)
            {
                if (board[row][col] != '.' && !rows.add(board[row][col]))
                    return false;
            }
        }
        return true;
    }
    
    public boolean colValids(char[][] board)
    {
        for(int row = 0; row < board.length; row++)
        {
            HashSet<Character> cols = new HashSet<Character>();
            for(int col = 0; col < 9; col++)
            {
                if (board[col][row] != '.' && !cols.add(board[col][row]))
                    return false;
            }
        }
        return true;
    }
    
    public boolean squareValids(char[][] board)
    {
        for(int row = 0; row < board.length; row++)
        {
            HashSet<Character> squares = new HashSet<Character>();
            for(int col = 0; col < board.length; col++)
            {
                int rowIdx = 3 * (row/3);
                int colIdx = 3 * (row%3);
                if (board[rowIdx + col/3][colIdx + col%3] != '.' && !squares.add(board[rowIdx + col/3][colIdx + col%3]))
                    return false;
            }
        }
        
        return true;
    }
}